X16 Reference - Appendix F - 65C816 Processor.md

```diff
@@ -0,0 +1,2470 @@
+```
+
+Tests the operand against the Accumulator. The ALU does an AND
+operation internally, and The **n**, **v**, and **z** flags are set accordingly.
+The Accumulator is *not* modified after the operation.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### BMI
+
+**Branch on Minus**
+
+```text
+SYNTAX         MODE          HEX LEN CYCLES      FLAGS   
+BMI LABEL      rel8          30  2   2+t+t*e*p   ........ .
+```
+
+Jumps to the specified address when the Negative flag (**n**) is set.
+
+**n** is set when ALU operations result in a negative number, or when the high bit
+of an ALU operation is 1.
+
+A branch operation uses an 8 bit signed value internally, starting from the
+instruction after the branch. So the branch destination can be 126 bytes before
+or 128 bytes after the branch instruction.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### BNE
+
+**Branch on Not Equal.**
+
+```text
+SYNTAX         MODE          HEX LEN CYCLES      FLAGS   
+BNE LABEL      rel8          D0  2   2+t+t*e*p   ........ .
+```
+
+Jumps to the target address when the Zero flag is 0. While this is most commonly
+used after a compare (CMP) operation, it's also useful to test if a number is
+zero after a Load operation, or to test if a loop is complete after a DEC
+operation.
+
+A branch operation uses an 8 bit signed value internally, starting from the
+instruction after the branch. So the branch destination can be 126 bytes before
+or 128 bytes after the branch instruction.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### BPL
+
+**Branch on Plus**
+
+```text
+SYNTAX         MODE          HEX LEN CYCLES      FLAGS   
+BPL LABEL      rel8          10  2   2+t+t*e*p   ........ .
+```
+
+Jumps to the specified address when the Negative flag (**n**) is clear.
+
+**n** is clear when ALU operations result in a positive number, or when the high bit
+of an ALU operation is 0.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### BRA
+
+**Branch Always**
+
+```text
+SYNTAX         MODE          HEX LEN CYCLES      FLAGS   
+BRA LABEL      rel8          80  2   3+e*p       ........ .
+```
+
+Jumps to the specified address.
+
+A branch operation uses an 8 bit signed value internally, starting from the
+instruction after the branch. So the branch destination can be 126 bytes before
+or 128 bytes after the branch instruction.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### BRK
+
+**Break**
+
+```text
+SYNTAX         MODE          HEX LEN CYCLES      FLAGS   
+BRK            imp           00  1   8-e         ....di.. .
```

Comment:
It does, and so does COP. Both should be documented as two byte opcodes with a one byte immediate operand, they are also documented that way in the 65C816's datasheet.

Reply:
So I recently learned that the value pushed to the stack by JSR is actually one byte before the real return address; this means the RTS instruction increments the PC at the end of the RTS operation. I'm guessing this is an artifact of how the CPU internally updates the PC.

Does Interrupt/RTI work the same way? Is the PC address on the stack actually the byte after BRK? 

This would explain why BRK appears to have an operand byte: this would be a quirk of the sequence in which the PC is incremented, 

This should be reflected in the documentation, since it currently says the PC is incremented by 2 during a BRK.